diff --git a/fc-solve/source/handle_parsing.h b/fc-solve/source/handle_parsing.h
index 1c86538d..fac53d89 100644
--- a/fc-solve/source/handle_parsing.h
+++ b/fc-solve/source/handle_parsing.h
@@ -19,8 +19,8 @@ extern "C" {
 #endif
 
 #include "rinutils.h"
+#if 0
 #include "fcs_cl.h"
-
 enum
 {
     EXIT_AND_RETURN_0 = FCS_CMD_LINE_USER
@@ -73,12 +73,22 @@ static inline void *alloc_instance_and_parse(const int argc, char **const argv,
 
     return instance;
 }
+#else
+#include "fcs_user.h"
+#include "fc_solve__theme__obf.c"
+#endif
 
 static inline void *simple_alloc_and_parse(
     const int argc, char **const argv, int *const arg_ptr)
 {
+#if 0
     return alloc_instance_and_parse(
         argc, argv, arg_ptr, NULL, NULL, NULL, TRUE);
+#else
+    void *const instance = freecell_solver_user_alloc();
+    my_init_instance(instance);
+    return instance;
+#endif
 }
 
 static inline void print_flares_plan_error(void *const instance)
